src/codegeneration/InnerForOnTransformer.js

```diff
@@ -0,0 +1,230 @@
+  ReturnStatement
+} from '../syntax/trees/ParseTrees.js';
+import {
+  createArgumentList,
+  createAssignmentStatement,
+  createAssignmentExpression,
+  createCaseClause,
+  createDefaultClause,
+  createIdentifierExpression as id,
+  createNumberLiteral,
+  createSwitchStatement,
+  createThisExpression,
+  createVariableDeclaration,
+  createVariableDeclarationList,
+  createVariableStatement,
+  createVoid0
+} from './ParseTreeFactory.js';
+import {ARGUMENTS} from '../syntax/PredefinedName.js';
+import {VAR} from '../syntax/TokenType.js';
+import {VARIABLE_DECLARATION_LIST} from '../syntax/trees/ParseTreeType.js';
+
+export class InnerForOnTransformer extends ParseTreeTransformer {
+
+  constructor(idGenerator, labelSet) {
+    this.idGenerator_ = idGenerator;
+    this.inLoop_ = 0;
+    this.inBreakble_ = 0;
+    this.variableDeclarations_ = [];
+    this.extractedStatements_ = [];
+    this.labelSet_ = labelSet;
+    this.labelledStatements_ = {};
+    this.observer_ = id(this.idGenerator_.generateUniqueIdentifier());
+    this.result_ = id(this.idGenerator_.generateUniqueIdentifier());
+    this.parentLabels_ = {};
+    this.labelSet_.forEach((tree) =>
+      this.parentLabels_[tree.name.value] = true
+    );
+  }
+
+  transform(tree) {
+    var value = id(this.idGenerator_.generateUniqueIdentifier());
+    var assignment;
+    if (tree.initializer.type === VARIABLE_DECLARATION_LIST) {
+      // {var,let} initializer = $value;
+      assignment = createVariableStatement(
+          tree.initializer.declarationType,
+          tree.initializer.declarations[0].lvalue, value);
+    } else {
+      assignment = parseStatement `
+          ${tree.initializer} = ${value};`;
+    }
+
+    var body = new Block(tree.body.location, [assignment].concat(tree.body.statements));
+    var body = this.transformAny(body);
+    body = alphaRenameThisAndArguments(this, body);
+
+    // var $result = undefined
+    this.variableDeclarations_.push(
+        createVariableDeclaration(this.result_, createVoid0()));
+
+    var caseClauses = this.extractedStatements_.map(
+        (statement, index) => createCaseClause(
+            createNumberLiteral(index), [statement])
+    );
+    caseClauses.push(createCaseClause(createVoid0(), [
+        new ContinueStatement(null, null)]));
+    caseClauses.push(createDefaultClause(parseStatements `
+        return ${this.result_}.v;`));
+
+    var switchStatement = createSwitchStatement(this.result_, caseClauses);
+    var statement = parseStatement `
+        do {
+          ${createVariableStatement(
+            createVariableDeclarationList(VAR, this.variableDeclarations_))}
+            await $traceurRuntime.observeForEach(
+              ${tree.observable}[$traceurRuntime.toProperty(Symbol.observer)].
+                  bind(${tree.observable}),
+              async function (${value}) {
+                var ${this.observer_} = this;
+                ${body}
+              });
+          ${switchStatement}
```

Comment:
In other words, the work should not happen in this PR but we should clean this up in the future.